### PR TITLE
ggml-opencl, llama: using reserve() if count already known

### PR DESCRIPTION
**Pull request by GermanAizek**
It affects a lot `ggml_cl_mul_mat_q_f32` function.

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -6116,6 +6116,7 @@ static bool llm_load_tensors(
                 mlock_buf->init   (ggml_backend_buffer_get_base(buf));
                 mlock_buf->grow_to(ggml_backend_buffer_get_size(buf));
             }
+            bufs.reserve(ml.files.size());
```
Comment (ggerganov) [inline llama.cpp:6119]: Already reserved on line 6060

File: ggml-opencl.cpp [hunk 2]
```diff
@@ -1835,7 +1835,10 @@ static void ggml_cl_mul_mat_q_f32(const ggml_tensor * src0, const ggml_tensor *
                     CL_CHECK(clEnqueueNDRangeKernel(queue, *to_fp32_cl, 1, &offset, &global, local > 0 ? &local : NULL, events.size(), !events.empty() ? events.data() : NULL, NULL));
                 }
 
-                for (int64_t i12 = i02 * r2, e12 = i12 + r2; i12 < e12; i12++) {
+                int64_t i12 = i02 * r2;
+                int64_t e12 = i12 + r2;
+                events.reserve(e12 - i12);
+                while (i12 < e12) {
```
Comment (ggerganov) [inline ggml-opencl.cpp:1841]: Better to keep the `for` loop

Comment (GermanAizek) [reply ggml-opencl.cpp]: fix it https://github.com/ggerganov/llama.cpp/pull/7272/commits/4ee29e5e1caf29e1bc7b094226faa890ae0e98d6

File: llama.cpp (hunk 1 above)
Comment (GermanAizek) [reply llama.cpp]: fix it https://github.com/ggerganov/llama.cpp/pull/7272/commits/4ee29e5e1caf29e1bc7b094226faa890ae0e98d6

File: ggml-opencl.cpp [hunk 3]
```diff
@@ -1835,7 +1835,10 @@ static void ggml_cl_mul_mat_q_f32(const ggml_tensor * src0, const ggml_tensor *
                     CL_CHECK(clEnqueueNDRangeKernel(queue, *to_fp32_cl, 1, &offset, &global, local > 0 ? &local : NULL, events.size(), !events.empty() ? events.data() : NULL, NULL));
                 }
 
-                for (int64_t i12 = i02 * r2, e12 = i12 + r2; i12 < e12; i12++) {
+                int64_t i12 = i02 * r2;
+                int64_t e12 = i12 + r2;
+                events.reserve(e12 - i12);
```
Comment (cebtenzzre) [inline ggml-opencl.cpp:1840]: For future reference: `events` is cleared at the end of this inner loop, so its actual maximum capacity is `3`. Even ignoring the `clear()`, `reserve()` does not grow the vector by the specified amount, it increases the capacity *to* the specified amount&mdash;so you would need to reserve `events.size() + e12 - i12` instead, if you were to even bother.

Luckily, this file is gone now, so this particular instance doesn't matter. But we should be more careful going forward.

Comment (GermanAizek) [reply ggml-opencl.cpp:1840]: @cebtenzzre, good catch. More reviewers there are, lower chance making a mistake, you're right.

**Pull request merged**